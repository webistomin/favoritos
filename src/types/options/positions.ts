export enum IFavoritosPositions {
  TOP_LEFT = 'top-left',
  TOP_RIGHT = 'top-right',
  BOTTOM_LEFT = 'bottom-left',
  BOTTOM_RIGHT = 'bottom-right',
}
