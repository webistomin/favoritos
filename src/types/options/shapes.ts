export enum IFavoritosShape {
  CIRCLE = 'circle',
  RECT = 'rectangle',
}
