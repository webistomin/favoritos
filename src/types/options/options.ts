import { IFavoritosPositions } from './positions';
import { IFavoritosShapes } from './shapes';

export interface IFavoritosOption {
  icon?: {
    iconSelector?: string;
    backgroundColor?: string | string[];
    shape?: IFavoritosShapes;
    lineWidth?: number;
    width?: number;
    height?: number;
  };
  badge?: {
    fontSize?: number;
    fontFamily?: string;
    backgroundColor?: string | string[];
    color?: string;
    position?: IFavoritosPositions;
    shape?: IFavoritosShapes;
    minWidth?: number;
    minHeight?: number;
  };
  debug?: {
    enabled?: boolean;
    debugSelector?: string;
  };
}
