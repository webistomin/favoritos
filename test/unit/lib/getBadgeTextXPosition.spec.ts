import Favoritos from '../../../src/favoritos';
import { FAVORITOS_POSITIONS } from '../../../src/types/options/positions';

describe('Favoritos: getBadgeTextXPosition method', () => {
  beforeEach(() => {
    document.head.innerHTML = `<link rel="shortcut icon" href="#">`;
  });

  it('Return correct X coord for number and "top-left" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.TOP_LEFT,
      },
    });
    lib['badgeContent'] = 1;
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for string and "top-left" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.TOP_LEFT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for number and "top-right" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.TOP_RIGHT,
      },
    });
    lib['badgeContent'] = 1;
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for string and "top-right" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.TOP_RIGHT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for number and "bottom-right" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.BOTTOM_RIGHT,
      },
    });
    lib['badgeContent'] = 1;
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for string and "bottom-right" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.BOTTOM_RIGHT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for number and "bottom-left" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.BOTTOM_LEFT,
      },
    });
    lib['badgeContent'] = 1;
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord for string and "bottom-left" position', () => {
    const lib = new Favoritos({
      badge: {
        position: FAVORITOS_POSITIONS.BOTTOM_LEFT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](10)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord if text width more than badge min width', () => {
    const lib = new Favoritos({
      badge: {
        minWidth: 22,
        position: FAVORITOS_POSITIONS.BOTTOM_RIGHT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](30)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](30)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('Return correct X coord if text width more than badge max width', () => {
    const lib = new Favoritos({
      icon: {
        width: 20,
      },
      badge: {
        minWidth: 22,
        position: FAVORITOS_POSITIONS.BOTTOM_RIGHT,
      },
    });
    lib['badgeContent'] = 'abc';
    expect(() => lib['getBadgeTextXPosition'](30)).not.toThrow(TypeError);
    expect(() => lib['getBadgeTextXPosition'](30)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });
});
