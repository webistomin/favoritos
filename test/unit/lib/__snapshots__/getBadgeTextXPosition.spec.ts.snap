// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for number and "bottom-left" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for number and "bottom-right" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for number and "top-left" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for number and "top-right" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for string and "bottom-left" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for string and "bottom-right" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for string and "top-left" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord for string and "top-right" position 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord if text width more than badge max width 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: getBadgeTextXPosition method Return correct X coord if text width more than badge min width 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;
